[DC-310] Update how we access table cells in UI tests

slucasbroad opened the pull request:
Determining the column index via the header information and “scanning” the rows of a table to find the correct data we are looking for in a given column.

slucasbroad committed:
adding 'aria-rowindex' and 'aria-colindex' to `SimpleTable()` and helper functions in `intergration-utils.js` to get them.

cahrens commented:
@slucasbroad welcome to terra-ui! I see this is your first PR.

It looks like you are still working on this, so I've made it a draft for now. When the PR is ready for review, please convert it to non-draft.

slucasbroad committed:
Replace helperfunctions finding by row and column with `findTableTextWithinColumn` and `getTableColIndex`.
Added @aria-colindex and @aria-rowindex to `simpleTable()` table.js

slucasbroad committed:
Removed all refrences to row and colum from `preview-dataset.js` and `request-access.js`

slucasbroad committed:
less eslint warnings

slucasbroad committed:
less eslint warnings

slucasbroad committed:
less eslint warnings

slucasbroad committed:
less eslint warnings

slucasbroad committed:
less eslint warnings

slucasbroad committed:
less eslint warnings

slucasbroad committed:
Rename Columntext to ColumnHeader

slucasbroad committed:
remove row/column from `catalog-utils.js`

slucasbroad committed:
fixed testing TimeoutError from `linkDataToWorkspace()` in catalog-utils.js calling a Controlled dataset.

petesantos commented on integration-tests/utils/catalog-utils.js:
```diff
@@ -10,12 +10,13 @@ const eitherThrow = (testFailure, { cleanupFailure, cleanupMessage }) => {
   }
 }
-
+//Transcriptomic and clonal characterization of T cells in the human central nervous system.
```
Remove comment
```suggestion
```

slucasbroad committed:
Update integration-tests/utils/catalog-utils.js

Co-authored-by: Pete Santos <50371603+petesantos@users.noreply.github.com>

cahrens commented:
@petesantos do you want to take over the review of this since you likely have more context?

cahrens commented:
@petesantos do you want to take over the review of this since you likely have more context?

petesantos commented on integration-tests/utils/catalog-utils.js:
```diff
@@ -10,12 +10,12 @@ const eitherThrow = (testFailure, { cleanupFailure, cleanupMessage }) => {
     throw new Error(`${cleanupMessage}: ${cleanupFailure.message}`)
   }
 }
-
```
nit: let's keep the spacing between the functions

fboulnois commented on src/components/table.js:
```diff
@@ -593,17 +595,19 @@ export const SimpleTable = ({
             return div({
               key,
               role: 'cell',
+              'aria-rowindex': i + 2,
```
Is the row index two because there's one row for the header followed by the data? If so, could we add that as a comment?
```suggestion
              'aria-rowindex': i + 2, // The first row of data is 2
```

fboulnois commented on integration-tests/utils/integration-utils.js:
```diff
@@ -315,5 +321,7 @@ module.exports = {
-  openError
+  openError,
+  getTableColIndex,
+  findTableTextWithinColumn
```
I see the other items are in alphabetical order, does this section need to follow this style?

fboulnois commented on integration-tests/tests/preview-dataset.js:
```diff
@@ -16,13 +16,10 @@ const testPreviewDatasetFn = withUserToken(async ({ testUrl, page, token }) => {
-  await findTableCellText(page, getTableCellPath(previewTableName, 2, 2), 'male')
-  await findTableCellText(page, getTableCellPath(previewTableName, 2, 3), '36')
+  const tableName = 'Participant Preview Data'
+  await findTableTextWithinColumn(page, { tableName, columnHeader: 'participant_id', textContains: 'participant1' })
```
I think it's useful to search by column name, but should check to ensure that a particular row of the table has the right data? For example, perhaps we can select the row by a column value (`participant_id` is `2`), and then ensure that row 2 indeed contains `2`, `female`, `33`:

```
participant_id | biological_sex | age
---------------+----------------+----
1              | male           | 26
2              | female         | 33
```

fboulnois reviewed (commented):
Overall the functionality looks reasonable to me, a few questions:

slucasbroad committed:
Update src/components/table.js

Co-authored-by: fboulnois <fboulnois@users.noreply.github.com>

slucasbroad committed:
Alphabetize new additions

slucasbroad committed:
Update preview-dataset.js

How about this for a test? Not sure if this actually tests for what you are asking:
`await getTableCellPath( page, { row: 2 , column: getTableColIndex(page, {tableName, columnHeader: 'age' }) })`

slucasbroad committed:
Update preview-dataset.js

slucasbroad committed:
Update preview-dataset.js

slucasbroad committed:
Update preview-dataset.js

slucasbroad committed:
Update preview-dataset.js

slucasbroad committed:
Update preview-dataset.js

slucasbroad committed:
create `assertRowMatches()` to test if a row with the given content exist a given table.

slucasbroad committed:
create `assertRowMatches()` to test if a row with the given content exist a given table.

slucasbroad committed:
create `assertRowMatches()` to test if a row with the given content exist a given table.

slucasbroad committed:
create `assertRowMatches()` to test if a row with the given content exist a given table.

slucasbroad committed:
simplifed `assertRowMatches()`

slucasbroad committed:
simplifed `assertRowMatches()`

slucasbroad committed:
simplifed `assertRowMatches()`

petesantos commented on integration-tests/utils/integration-utils.js:
```diff
@@ -66,17 +66,38 @@ const getTableCellPath = (tableName, row, column) => {
+  return Promise.all(_.map(findTextInColumn, expectedColumnValues))
+}
+
+const getTableRowIndex = async (page, { tableName, columnHeader, textContains, isDescendant = false }) => {
```
nit: Can you move this function above `assertRowMatches`. Since the assert calls this function, we try to have the functions defined before their usage

petesantos commented on integration-tests/utils/integration-utils.js:
```diff
@@ -66,17 +66,38 @@ const getTableCellPath = (tableName, row, column) => {
 
-const clickTableCell = async (page, tableName, row, column, options) => {
-  const tableCellPath = getTableCellPath(tableName, row, column)
+const assertRowMatches = async (page, { tableName, expectedColumnValues, withKey: { column, textContains } }) => {
```
nit: Could you rename this to `assertRowHas`?

```suggestion
const assertRowHas = async (page, { tableName, expectedColumnValues, withKey: { column, textContains } }) => {
```

petesantos reviewed (commented):
Looks good! A few minor changes

petesantos commented on integration-tests/utils/integration-utils.js:
```diff
@@ -66,17 +66,38 @@ const getTableCellPath = (tableName, row, column) => {
+    return await findElement(page, xPath, { timeout: 5000 })
+  }
+
+  return Promise.all(_.map(findTextInColumn, expectedColumnValues))
```
Since this is an assertion, let's have this not return anything. It will throw if one of the async calls fail
```suggestion
await Promise.all(_.map(findTextInColumn, expectedColumnValues))
```

slucasbroad committed:
Update integration-tests/utils/integration-utils.js

Co-authored-by: Pete Santos <50371603+petesantos@users.noreply.github.com>

slucasbroad committed:
Update integration-utils.js

Refactor `assertRowMatches` to `assertRowHas`

slucasbroad committed:
Update preview-dataset.js

Refactor `assertRowMatches` to `assertRowHas`

slucasbroad committed:
Update integration-utils.js

slucasbroad committed:
Update integration-utils.js

slucasbroad committed:
Update catalog-utils.js

petesantos reviewed (approved):
Looks good, thanks for doing this!

s-rubenstein reviewed (approved):
LGTM

aweng98 commented on integration-tests/utils/integration-utils.js:
```diff
@@ -66,17 +66,38 @@ const getTableCellPath = (tableName, row, column) => {
-const getTableHeaderPath = (tableName, column) => {
-  return `//*[@role="table" and @aria-label="${tableName}"]//*[@role="row"][1]//*[@role="columnheader"][${column}]`
+const getTableColIndex = async (page, { tableName, columnHeader }) => {
+  const colHeaderNode = await findElement(page, `//*[@role="table" and @aria-label="${tableName}"]//*[@role="columnheader" and @aria-colindex and contains(normalize-space(.),"${columnHeader}")]`)
```
`contains(normalize-space(.),"${columnHeader}")` could be problematic if more than one header text contain `columnHeader` string. like `Name` and `User Name`. Why not use `text()=columnHeader`?

aweng98 commented on integration-tests/utils/integration-utils.js:
```diff
@@ -66,17 +66,38 @@ const getTableCellPath = (tableName, row, column) => {
-const findTableCellText = async (page, path, textContains, options) => {
-  const xpath = `${path}[contains(normalize-space(.),"${textContains}")]`
-  return (await page.waitForXPath(xpath, options))
+const getTableTextWithinColumn = async (page, { tableName, columnHeader, textContains, isDescendant = false }) => {
```
Function name suggest it returns text. But it returns a xpath.

aweng98 replied:
How do you handle finding if cell contains non-text element like an icon or image?

aweng98 replied:
Example of [findCellByRowValue](https://github.com/all-of-us/workbench/blob/f84663918a98a53ddc626811262650d99f6c7c63/e2e/app/component/table.ts#L113)

aweng98 reviewed (commented):
Few comments that could make functions more versatile for your consideration.
